File: android/core/protocol/src/main/kotlin/com/cayatur/winbridge/protocol/Channels.kt
package com.cayatur.winbridge.protocol

/**
 * Which queue a frame goes on. One socket carries control, media and bulk
 * traffic, so a strict priority queue in front of the writer is what keeps a
 * touch event from waiting behind a 48 KB tile batch or a file chunk.
 */
enum class SendLane { CONTROL, MEDIA, BULK }

object MediaKind {
    const val VIDEO: Byte = 1
    const val AUDIO: Byte = 2
}

/**
 * Stream identifiers are fixed rather than negotiated: both ends must agree on
 * what "stream 3" means before any control message about it can be exchanged.
 */
object StreamIds {
    const val PC_SCREEN: Byte = 0
    const val PHONE_SCREEN: Byte = 1
    const val PC_AUDIO: Byte = 2
    const val PHONE_AUDIO: Byte = 3
    const val PC_MIC: Byte = 4
    const val PHONE_MIC: Byte = 5

    fun name(id: Byte): String = when (id) {
        PC_SCREEN -> "pc.screen"
        PHONE_SCREEN -> "phone.screen"
        PC_AUDIO -> "pc.audio"
        PHONE_AUDIO -> "phone.audio"
        PC_MIC -> "pc.mic"
        PHONE_MIC -> "phone.mic"
        else -> "stream$id"
    }

    fun fromName(name: String): Byte = when (name) {
        "pc.screen" -> PC_SCREEN
        "phone.screen" -> PHONE_SCREEN
        "pc.audio" -> PC_AUDIO
        "phone.audio" -> PHONE_AUDIO
        "pc.mic" -> PC_MIC
        "phone.mic" -> PHONE_MIC
        else -> -1
    }
}

object MediaFlags {
    const val NONE = 0
    /** Every tile of the frame is present; the receiver may reset its canvas. */
    const val KEYFRAME = 1
    /** Last packet of this frame — the receiver may present now. */
    const val END_OF_FRAME = 2
}

/**
 * One real-time packet: 11-byte header then payload. Deliberately not JSON —
 * at sixty packets a second both the parse cost and the byte overhead matter.
 */
class MediaPacket(
    val kind: Byte,
    val stream: Byte,
    val seq: Int,
    val timestampMs: Int,
    val flags: Int,
    val payload: ByteArray,
    val payloadOffset: Int = 0,
    val payloadLength: Int = payload.size,
) {
    fun toBytes(): ByteArray {
        val body = ByteArray(1 + HEADER_SIZE + payloadLength)
        body[0] = InnerType.MEDIA.code
        body[1] = kind
        body[2] = stream
        writeIntBE(body, 3, seq)
        writeIntBE(body, 7, timestampMs)
        body[11] = flags.toByte()
        payload.copyInto(body, 1 + HEADER_SIZE, payloadOffset, payloadOffset + payloadLength)
        return body
    }

    val isKeyframe: Boolean get() = flags and MediaFlags.KEYFRAME != 0
    val isEndOfFrame: Boolean get() = flags and MediaFlags.END_OF_FRAME != 0

    companion object {
        const val HEADER_SIZE = 11

        /** @param plaintext the decrypted inner frame, including the leading inner-type byte. */
        fun parse(plaintext: ByteArray): MediaPacket {
            if (plaintext.size < 1 + HEADER_SIZE) throw ProtocolException("truncated media packet")
            return MediaPacket(
                kind = plaintext[1],
                stream = plaintext[2],
                seq = readIntBE(plaintext, 3),
                timestampMs = readIntBE(plaintext, 7),
                flags = plaintext[11].toInt() and 0xFF,
                payload = plaintext,
                payloadOffset = 1 + HEADER_SIZE,
                payloadLength = plaintext.size - 1 - HEADER_SIZE,
            )
        }
    }
}

object XferFlags {
    const val NONE = 0
    const val LAST = 1
}

/** One chunk of a file transfer. */
class XferChunk(
    val transferId: Int,
    val seq: Int,
    val flags: Int,
    val data: ByteArray,
    val dataOffset: Int = 0,
    val dataLength: Int = data.size,
) {
    fun toBytes(): ByteArray {
        val body = ByteArray(1 + HEADER_SIZE + dataLength)
        body[0] = InnerType.XFER.code
        writeIntBE(body, 1, transferId)
        writeIntBE(body, 5, seq)
        body[9] = flags.toByte()
        data.copyInto(body, 1 + HEADER_SIZE, dataOffset, dataOffset + dataLength)
        return body
    }

    val isLast: Boolean get() = flags and XferFlags.LAST != 0

    companion object {
        const val HEADER_SIZE = 9

        fun parse(plaintext: ByteArray): XferChunk {
            if (plaintext.size < 1 + HEADER_SIZE) throw ProtocolException("truncated xfer chunk")
            return XferChunk(
                transferId = readIntBE(plaintext, 1),
                seq = readIntBE(plaintext, 5),
                flags = plaintext[9].toInt() and 0xFF,
                data = plaintext,
                dataOffset = 1 + HEADER_SIZE,
                dataLength = plaintext.size - 1 - HEADER_SIZE,
            )
        }
    }
}

/**
 * A screen frame is a list of changed tiles: `index:u16, len:u32, jpeg[len]`,
 * repeated. Only tiles whose content hash moved are sent, which is what makes an
 * intra-only codec affordable — and unlike an inter-frame codec there is no
 * reference-frame latency to pay on either side.
 */
object TileCodec {
    /** Keep one packet small enough that it never becomes a visible stall. */
    const val MAX_PACKET_BYTES = 48 * 1024

    fun writeTile(destination: java.io.ByteArrayOutputStream, index: Int, jpeg: ByteArray, length: Int = jpeg.size) {
        destination.write((index ushr 8) and 0xFF)
        destination.write(index and 0xFF)
        destination.write((length ushr 24) and 0xFF)
        destination.write((length ushr 16) and 0xFF)
        destination.write((length ushr 8) and 0xFF)
        destination.write(length and 0xFF)
        destination.write(jpeg, 0, length)
    }

    /** Calls [onTile] for each tile in a media payload. Slices are views, not copies. */
    inline fun forEachTile(
        payload: ByteArray,
        offset: Int,
        length: Int,
        onTile: (index: Int, data: ByteArray, dataOffset: Int, dataLength: Int) -> Unit,
    ) {
        var cursor = offset
        val end = offset + length
        while (cursor + 6 <= end) {
            val index = ((payload[cursor].toInt() and 0xFF) shl 8) or (payload[cursor + 1].toInt() and 0xFF)
            val size = readIntBE(payload, cursor + 2)
            cursor += 6
            if (size < 0 || size > end - cursor) return
            onTile(index, payload, cursor, size)
            cursor += size
        }
    }
}

internal fun writeIntBE(buffer: ByteArray, offset: Int, value: Int) {
    buffer[offset] = (value ushr 24).toByte()
    buffer[offset + 1] = (value ushr 16).toByte()
    buffer[offset + 2] = (value ushr 8).toByte()
    buffer[offset + 3] = value.toByte()
}

fun readIntBE(buffer: ByteArray, offset: Int): Int =
    ((buffer[offset].toInt() and 0xFF) shl 24) or
        ((buffer[offset + 1].toInt() and 0xFF) shl 16) or
        ((buffer[offset + 2].toInt() and 0xFF) shl 8) or
        (buffer[offset + 3].toInt() and 0xFF)

File: windows/src/WinBridge.Core/Protocol/Channels.cs
using System.Buffers.Binary;

namespace WinBridge.Core.Protocol;

/// <summary>
/// Which lane a frame is queued on. One socket carries control, media and bulk
/// traffic, so a strict priority queue in front of the writer is what keeps a
/// touch event from queueing behind a 60 KB screen tile batch or a file chunk.
/// </summary>
public enum SendLane
{
    /// <summary>Commands, input, heartbeat, state. Never dropped, always first.</summary>
    Control = 0,
    /// <summary>Screen tiles and audio. Dropped under pressure — stale is worse than absent.</summary>
    Media = 1,
    /// <summary>File chunks. Never dropped, but always last and flow-controlled.</summary>
    Bulk = 2,
}

public static class MediaKind
{
    public const byte Video = 1;
    public const byte Audio = 2;
}

/// <summary>
/// Stream identifiers are fixed rather than negotiated: both ends need to agree
/// on what "stream 3" means before any control message can be exchanged about
/// it, and there are few enough of them that a registry buys nothing.
/// </summary>
public static class StreamIds
{
    public const byte PcScreen = 0;
    public const byte PhoneScreen = 1;
    public const byte PcAudio = 2;      // what the PC is playing (WASAPI loopback)
    public const byte PhoneAudio = 3;   // what the phone is playing (AudioPlaybackCapture)
    public const byte PcMic = 4;
    public const byte PhoneMic = 5;

    public static string Name(byte id) => id switch
    {
        PcScreen => "pc.screen",
        PhoneScreen => "phone.screen",
        PcAudio => "pc.audio",
        PhoneAudio => "phone.audio",
        PcMic => "pc.mic",
        PhoneMic => "phone.mic",
        _ => $"stream{id}",
    };

    public static byte FromName(string name) => name switch
    {
        "pc.screen" => PcScreen,
        "phone.screen" => PhoneScreen,
        "pc.audio" => PcAudio,
        "phone.audio" => PhoneAudio,
        "pc.mic" => PcMic,
        "phone.mic" => PhoneMic,
        _ => 0xFF,
    };
}

[Flags]
public enum MediaFlags : byte
{
    None = 0,
    /// <summary>Every tile of the frame is present; the receiver may reset its canvas.</summary>
    Keyframe = 1,
    /// <summary>Last packet of this frame — the receiver may present now.</summary>
    EndOfFrame = 2,
}

/// <summary>
/// One real-time packet. Header is 11 bytes: kind, stream, seq, timestamp, flags.
/// Deliberately not JSON — at 60 packets a second the parse cost and the byte
/// overhead both matter.
/// </summary>
public readonly record struct MediaPacket(
    byte Kind, byte Stream, uint Seq, uint TimestampMs, MediaFlags Flags, ReadOnlyMemory<byte> Payload)
{
    public const int HeaderSize = 11;

    public byte[] ToBytes()
    {
        var body = new byte[1 + HeaderSize + Payload.Length];
        body[0] = (byte)InnerType.Media;
        body[1] = Kind;
        body[2] = Stream;
        BinaryPrimitives.WriteUInt32BigEndian(body.AsSpan(3), Seq);
        BinaryPrimitives.WriteUInt32BigEndian(body.AsSpan(7), TimestampMs);
        body[11] = (byte)Flags;
        Payload.Span.CopyTo(body.AsSpan(1 + HeaderSize));
        return body;
    }

    /// <param name="plaintext">The decrypted inner frame, including the leading inner type byte.</param>
    public static MediaPacket Parse(ReadOnlyMemory<byte> plaintext)
    {
        if (plaintext.Length < 1 + HeaderSize) throw new ProtocolException("truncated media packet");
        var span = plaintext.Span;
        return new MediaPacket(
            span[1],
            span[2],
            BinaryPrimitives.ReadUInt32BigEndian(span[3..]),
            BinaryPrimitives.ReadUInt32BigEndian(span[7..]),
            (MediaFlags)span[11],
            plaintext[(1 + HeaderSize)..]);
    }
}

[Flags]
public enum XferFlags : byte
{
    None = 0,
    Last = 1,
}

/// <summary>One chunk of a file transfer.</summary>
public readonly record struct XferChunk(uint TransferId, uint Seq, XferFlags Flags, ReadOnlyMemory<byte> Data)
{
    public const int HeaderSize = 9;

    public byte[] ToBytes()
    {
        var body = new byte[1 + HeaderSize + Data.Length];
        body[0] = (byte)InnerType.Xfer;
        BinaryPrimitives.WriteUInt32BigEndian(body.AsSpan(1), TransferId);
        BinaryPrimitives.WriteUInt32BigEndian(body.AsSpan(5), Seq);
        body[9] = (byte)Flags;
        Data.Span.CopyTo(body.AsSpan(1 + HeaderSize));
        return body;
    }

    public static XferChunk Parse(ReadOnlyMemory<byte> plaintext)
    {
        if (plaintext.Length < 1 + HeaderSize) throw new ProtocolException("truncated xfer chunk");
        var span = plaintext.Span;
        return new XferChunk(
            BinaryPrimitives.ReadUInt32BigEndian(span[1..]),
            BinaryPrimitives.ReadUInt32BigEndian(span[5..]),
            (XferFlags)span[9],
            plaintext[(1 + HeaderSize)..]);
    }
}

/// <summary>
/// A screen frame is a list of changed tiles. Only the tiles whose content hash
/// moved are sent, which is what makes an intra-only codec affordable: a desktop
/// with a blinking cursor costs a few hundred bytes a frame, and there is no
/// reference-frame latency to pay on either side.
/// </summary>
public static class TileCodec
{
    /// <summary>Keep a single packet comfortably under the MTU-amplified cost of a stall.</summary>
    public const int MaxPacketBytes = 48 * 1024;

    public static void WriteTile(Stream destination, ushort index, ReadOnlySpan<byte> jpeg)
    {
        Span<byte> header = stackalloc byte[6];
        BinaryPrimitives.WriteUInt16BigEndian(header, index);
        BinaryPrimitives.WriteUInt32BigEndian(header[2..], (uint)jpeg.Length);
        destination.Write(header);
        destination.Write(jpeg);
    }

    public static IEnumerable<(ushort Index, ReadOnlyMemory<byte> Jpeg)> ReadTiles(ReadOnlyMemory<byte> payload)
    {
        int offset = 0;
        while (offset + 6 <= payload.Length)
        {
            ushort index = BinaryPrimitives.ReadUInt16BigEndian(payload.Span[offset..]);
            uint length = BinaryPrimitives.ReadUInt32BigEndian(payload.Span[(offset + 2)..]);
            offset += 6;
            if (length > (uint)(payload.Length - offset)) yield break;
            yield return (index, payload.Slice(offset, (int)length));
            offset += (int)length;
        }
    }
}
